tsdate/prior.py

```diff
@@ -1011,9 +1139,11 @@ def build_grid(
 
     :param TreeSequence tree_sequence: The input :class:`tskit.TreeSequence`, treated as
         undated.
-    :param float Ne: The estimated (diploid) effective population size: must be
-        specified. Using standard (unscaled) values for ``Ne`` results in a prior where
-        times are measures in generations.
+    :param float population_size: The estimated (diploid) effective population
+        size: must be specified. May be a single value, or a two-column array with
+        epoch breakpoints and effective population sizes. Using standard (unscaled)
```

Comment:
How do the pop sizes map to the breakpoints? Do we require the same number of rows in the 2 columns, and if so, is one ignored?

It would be nice to be able to specify a continuous Ne function here too, instead of a piecewise set of points, wouldn't it? Would this be any technically more difficult as a rescaling method? I don't think it should...

Reply:
It does turn out to be technically more difficult, because the user is passing in a continuous population size function $n(t)$ that is parameterized by time in generations. Instead, we need a population size function that is parameterized in coalescent units, because we're converting from coalescent timescale to natural timescale. This won't generally have a closed form. It can be calculated numerically, but it involves optimization where the objective function uses quadrature, and it's pretty easy for this to futz up numerically. (Unless I'm missing something obvious).

On the other hand, if the population size function is piecewise constant, then it's really easy to convert back-and-forth. Ultimately, I think the thing to do (to allow continuous $n(t)$) is to approximate it with piecewise constant, and use a nested quadrature rule (Romberg or something similar) to get the approximation to within a desired accuracy. But, this'll be somewhat involved to implement, and I think it's a good first step to just get the piecewise constant case working first.

Reply:
Right, I see. I think this is an integration thing, right? We need the integral of the rate function to convert to successive time points. So I guess we could require the integral to be passed, rather than the actual population size function?

As you say, though, it's good to get the piecewise version working first. I'm always a bit scared of these piecewise things: the discrete jumps between time boundaries feels very wrong to me!

Reply:
Hm, I think that'd be a pretty stringent requirement for the user. Except for a few specific cases (piecewise constant, exponential) I don't think the integral would have a closed form. It seems better to support these explictly (e.g. add a "growth_rate" array for piecewise-exponential models) or to do the numerics under the hood by sampling the continous function on successively finer grids until the time rescaling converges. But let's cross that bridge later ...

In my experience, as long as the piecewise constant history is a fine enough discretization, it'll do a very good job of approximating a continuous population size history. After all, a discrete-time Wright-Fisher model is effectively a piecewise-constant one, with each piece the length of a generation.